File: iteration2/src/course/Course.java
package iteration2.src.course;
import iteration2.src.Department;
import iteration2.src.RandomNumberGenerator;
import iteration2.src.human.Assistant;
import iteration2.src.human.Grade;
import iteration2.src.human.Lecturer;
import iteration2.src.human.Student;
import iteration2.src.input_output.Logger;

import java.util.ArrayList;
import java.util.List;

public abstract class Course {
    private static final int minQuota = 40;
    private static final int maxQuota = 80;

    protected String code;
    protected String name;
    protected int credits;
    protected int ects;
    protected int theoreticalHours;
    protected int appliedHours;
    protected Grade firstYearToTake;
    protected Season firstSeasonToTake;
    protected int quota;

    protected List<Course> prerequisites = new ArrayList<>();
    protected List<Lecturer> lecturers = new ArrayList<>();
    protected List<Assistant> assistants = new ArrayList<>();
    protected List<CourseSection> courseSections = new ArrayList<>();
    protected List<LabSection> labSections = new ArrayList<>();

    public Course(String code, String name, int credits, int theoreticalHours, int appliedHours,
                  Grade firstYearToTake,Season firstSeasonToTake, List<Lecturer> lecturers, List<Assistant> assistants){
        this.code = code;
        this.name = name;
        this.credits = credits;
        this.ects = credits;
        this.theoreticalHours = theoreticalHours;
        this.appliedHours = appliedHours;
        this.firstYearToTake = firstYearToTake;
        this.firstSeasonToTake = firstSeasonToTake;

        if(theoreticalHours > 0)
            this.lecturers = lecturers;

        if(appliedHours > 0)
            this.assistants = assistants;

        this.quota = RandomNumberGenerator.randomIntegerBetween(minQuota,maxQuota + 1);

        //Each and every semester, at least one section of all mandatory courses should be registrable without any collision
    }

    public void addPrerequisite(Course prerequisite){
        prerequisites.add(prerequisite);
    }
    public CourseSection addCourseSection(long schedule){
        if(theoreticalHours == 0)
            return null;

        int randomIndex = RandomNumberGenerator.randomIntegerBetween(0, lecturers.size());
        Lecturer lecturer = lecturers.get(randomIndex);
        String sectionCode = String.valueOf(courseSections.size() + 1);

        CourseSection newSection = new CourseSection(this,sectionCode,schedule,lecturer);
        courseSections.add(newSection);

        return newSection;
    }

    public LabSection addLabSection(long schedule){
        if(appliedHours == 0)
            return null;

        int randomIndex = RandomNumberGenerator.randomIntegerBetween(0, assistants.size());
        Assistant assistant = assistants.get(randomIndex);

        StringBuilder sectionCode = new StringBuilder();
        sectionCode.append(courseSections.size());
        sectionCode.append('.');
        sectionCode.append(labSections.size() + 1);

        LabSection newSection = new LabSection(this,sectionCode.toString(),schedule,assistant);
        labSections.add(newSection);

        return newSection;
    }

    public boolean isStudentGradeRequirementMet(Student student){
        return getCourseSemester() <= student.getStudentSemester();
    }

    public Boolean canStudentTakeCourse(Student student){
        return isStudentGradeRequirementMet(student) && !student.didStudentPass(this)
                && student.checkIfPrerequisitesArePassed(this);
    }


    public List<CourseSection> getAvailableCourseSections(){
        List<CourseSection> sections = new ArrayList<>();

        for (CourseSection s: courseSections){
            if(!s.isSectionFull())
                sections.add(s);
        }

        return sections;
    }

    public List<Section> getAlternativeSections(Section section){
        List<Section> alternatives = new ArrayList<>();

        if(section instanceof CourseSection) {
            for (Section s : courseSections)
                if (!section.equals(s))
                    alternatives.add(s);
        }
        else{
            for (Section s : labSections)
                if(!section.equals(s))
                    alternatives.add(s);
        }
        return alternatives;
    }

    public List<LabSection> getAvailableLabSections(){
        List<LabSection> sections = new ArrayList<>();

        for(LabSection s: labSections){
            if(!s.isSectionFull())
                sections.add(s);
        }

        return sections;
    }

    public void requestNewCourseSection(){ }

    //Getters
    public abstract int getCoursePriority();
    public int getCourseSemester(){
        return 2 * firstYearToTake.getValue() + firstSeasonToTake.getValue();
    }
    public int getCredits() {
        return credits;
    }
    public String getCode() {
        return code;
    }
    public String getName(){
        return name;
    }
    public List<Course> getPrerequisites() {
        return prerequisites;
    }
    public int getTheoreticalHours() {
        return theoreticalHours;
    }
    public int getAppliedHours() {
        return appliedHours;
    }
    public Grade getFirstYearToTake() {
        return firstYearToTake;
    }
    public List<Lecturer> getLecturers() {
        return lecturers;
    }
    public Season getFirstSeasonToTake() {
        return firstSeasonToTake;
    }
    public List<Assistant> getAssistants() {
        return assistants;
    }

    public List<CourseSection> getCourseSections(){
        return courseSections;
    }
    public List<Section> getAllSections() {
        List<Section> allSections = new ArrayList<>(courseSections);
        allSections.addAll(labSections);
        return allSections;
    }
    public int getQuota() {
        return quota;
    }
}

File: iteration3/Course.py
import iteration2.src.math_helper
from iteration2.src.human import Assistant, Grade, Lecturer, Student
from iteration2.src.course_section import CourseSection
from iteration2.src.lab_section import LabSection
from typing import List

class Course:
    MIN_QUOTA = 40
    MAX_QUOTA = 80

    def __init__(self, code: str, name: str, credits: int, theoretical_hours: int, applied_hours: int,
                 first_year_to_take: Grade, first_season_to_take: 'Season', lecturers: List[Lecturer], assistants: List[Assistant]):
        self.code = code
        self.name = name
        self.credits = credits
        self.ects = credits
        self.theoretical_hours = theoretical_hours
        self.applied_hours = applied_hours
        self.first_year_to_take = first_year_to_take
        self.first_season_to_take = first_season_to_take

        if theoretical_hours > 0:
            self.lecturers = lecturers

        if applied_hours > 0:
            self.assistants = assistants

        self.quota = math_helper.random_integer_between(MIN_QUOTA, MAX_QUOTA + 1)

    def add_prerequisite(self, prerequisite: 'Course'):
        self.prerequisites.append(prerequisite)

    def add_course_section(self, schedule: int) -> CourseSection:
        if self.theoretical_hours == 0:
            return None

        random_index = math_helper.random_integer_between(0, len(self.lecturers))
        lecturer = self.lecturers[random_index]
        section_code = str(len(self.course_sections) + 1)

        new_section = CourseSection(self, section_code, schedule, lecturer)
        self.course_sections.append(new_section)

        return new_section

    def add_lab_section(self, schedule: int) -> 'LabSection':
        if self.applied_hours == 0:
            return None

        random_index = math_helper.random_integer_between(0, len(self.assistants))
        assistant = self.assistants[random_index]

        section_code = f'{len(self.course_sections)}.{len(self.lab_sections) + 1}'
        new_section = LabSection(self, section_code, schedule, assistant)
        self.lab_sections.append(new_section)

        return new_section

    def is_student_grade_requirement_met(self, student: Student) -> bool:
        return self.get_course_semester() <= student.get_student_semester()

    def can_student_take_course(self, student: Student) -> bool:
        return self.is_student_grade_requirement_met(student) and not student.did_student_pass(self) and student.check_if_prerequisites_are_passed(self)

    def get_available_course_sections(self) -> List[CourseSection]:
        sections = []
        return [s for s in self.course_sections if not s.is_section_full()]

    def get_alternative_sections(self, section: 'Section') -> List['Section']:
        alternatives = []

        if isinstance(section, CourseSection):
            for s in self.course_sections:
                if s != section and not s.is_section_full():
                    alternatives.append(s)
        else:
            for s in self.lab_sections:
                if s != section and not s.is_section_full():
                    alternatives.append(s)

        return alternatives

    def get_code(self) -> str:
        return self.code

    def get_name(self) -> str:
        return self.name

    def get_credits(self) -> int:
        return self.credits

    def get_ects(self) -> int:
        return self.ects

    def get_theoretical_hours(self) -> int:
        return self.theoretical_hours

    def get_applied_hours(self) -> int:
        return self.applied_hours

    def get_first_year_to_take(self) -> Grade:
        return self.first_year_to_take

    def get_first_season_to_take(self) -> 'Season':
        return self.first_season_to_take

    def get_quota(self) -> int:
        return self.quota

    def get_prerequisites(self) -> List['Course']:
        return self.prerequisites

    def get_lecturers(self) -> List[Lecturer]:
        return self.lecturers

    def get_assistants(self) -> List[Assistant]:
        return self.assistants

    def get_course_sections(self) -> List[CourseSection]:
        return self.course_sections

    def get_lab_sections(self) -> List[LabSection]:
        return self.lab_sections

